File: android-app/app/src/main/java/ke/co/aiccathedral/messenger/MainActivity.java
package ke.co.aiccathedral.messenger;

import android.Manifest;
import android.app.*;
import android.content.*;
import android.content.pm.PackageManager;
import android.os.*;
import android.telephony.*;
import android.view.*;
import android.widget.*;
import java.io.*;
import java.net.*;
import java.util.*;
import org.json.*;

public class MainActivity extends Activity {
    static MainActivity instance;
    static final String SB_URL="https://rmarersocrwzqhnbuygi.supabase.co";
    static final String SB_KEY="sb_publishable_mVPoh33HbSN0Vyo5qHyWQQ_zgX38BUo";
    static final int REQ_SMS=41;
    LinearLayout root,list; EditText message; Spinner simSpinner; TextView counter,summary; Button send;
    ArrayList<Contact> contacts=new ArrayList<>(); ArrayList<SubscriptionInfo> sims=new ArrayList<>();
    Handler handler=new Handler(Looper.getMainLooper()); String accessToken="";
    int sent=0,failed=0,delivered=0,total=0; boolean sending=false;

    @Override public void onCreate(Bundle b){
        super.onCreate(b); instance=this; showLogin();
        if(Build.VERSION.SDK_INT>=23 && checkSelfPermission(Manifest.permission.SEND_SMS)!=PackageManager.PERMISSION_GRANTED)
            requestPermissions(new String[]{Manifest.permission.SEND_SMS,Manifest.permission.READ_PHONE_STATE},REQ_SMS);
    }
    TextView tv(String s,int size){TextView t=new TextView(this);t.setText(s);t.setTextSize(size);t.setTextColor(0xFF242024);t.setPadding(0,8,0,8);return t;}
    EditText field(String hint){EditText e=new EditText(this);e.setHint(hint);e.setPadding(12,8,12,8);return e;}
    void base(){ScrollView sc=new ScrollView(this);root=new LinearLayout(this);root.setOrientation(LinearLayout.VERTICAL);root.setPadding(28,24,28,28);sc.addView(root);setContentView(sc);}
    void showLogin(){
        base(); TextView b=tv("AIC CATHEDRAL",28);b.setTextColor(0xFF7A1F3D);b.setTypeface(null,1);root.addView(b);
        root.addView(tv("School SMS Messenger",16)); root.addView(tv("Sign in with your ERP account to load parents and students.",14));
        EditText email=field("ERP email"); email.setInputType(33); root.addView(email);
        EditText pass=field("Password"); pass.setInputType(129); root.addView(pass);
        Button login=new Button(this);login.setText("SIGN IN");root.addView(login);
        TextView status=tv("",14);root.addView(status);
        login.setOnClickListener(v->{login.setEnabled(false);status.setText("Signing in...");new Thread(()->{
            try{
                String body="{\"email\":"+JSONObject.quote(email.getText().toString().trim())+",\"password\":"+JSONObject.quote(pass.getText().toString())+"}";
                JSONObject j=request(SB_URL+"/auth/v1/token?grant_type=password","POST",body,null);
                accessToken=j.getString("access_token");
                runOnUiThread(this::showMessenger);
            }catch(Exception e){runOnUiThread(()->{status.setText("Login failed: "+e.getMessage());login.setEnabled(true);});}
        }).start();});
    }
    void showMessenger(){
        base(); TextView brand=tv("AIC CATHEDRAL",26);brand.setTextColor(0xFF7A1F3D);brand.setTypeface(null,1);root.addView(brand);
        root.addView(tv("ERP Parent SMS",15));
        LinearLayout card=new LinearLayout(this);card.setOrientation(LinearLayout.VERTICAL);card.setPadding(20,18,20,18);card.setBackgroundColor(0xFFF7F1F4);root.addView(card);
        counter=tv("0 / 100 selected",22);counter.setTypeface(null,1);card.addView(counter);
        Button load=new Button(this);load.setText("Refresh parents from ERP");load.setOnClickListener(v->loadParents());card.addView(load);
        list=new LinearLayout(this);list.setOrientation(LinearLayout.VERTICAL);card.addView(list);
        card.addView(tv("Sending SIM",14));simSpinner=new Spinner(this);card.addView(simSpinner);loadSims();
        message=field("Type SMS message");message.setMinLines(4);message.setGravity(Gravity.TOP);card.addView(message);
        send=new Button(this);send.setText("SEND SMS TO SELECTED");send.setOnClickListener(v->startSending());card.addView(send);
        summary=tv("No message sent yet",16);root.addView(summary);
        Button all=new Button(this);all.setText("Select all loaded parents (max 100)");all.setOnClickListener(v->{for(Contact c:contacts)c.selected=true;refresh();});root.addView(all);
        Button clear=new Button(this);clear.setText("Clear selection");clear.setOnClickListener(v->{for(Contact c:contacts)c.selected=false;refresh();});root.addView(clear);
        loadParents();
    }
    void loadParents(){
        summary.setText("Loading active students and parents from ERP...");
        new Thread(()->{
            try{
                String sel="id,first_name,middle_name,last_name,admission_number,student_parents(parent_id,primary_guardian,parents(id,name,phone,status))";
                String url=SB_URL+"/rest/v1/students?select="+URLEncoder.encode(sel,"UTF-8")+"&status=eq.active&order=first_name.asc";
                JSONObject[] arrObj={}; JSONArray a=new JSONArray(requestText(url,"GET",null,accessToken));
                ArrayList<Contact> loaded=new ArrayList<>();
                for(int i=0;i<a.length();i++){
                    JSONObject s=a.getJSONObject(i);JSONArray links=s.optJSONArray("student_parents");Contact chosen=null;
                    if(links!=null) for(int k=0;k<links.length();k++){JSONObject l=links.optJSONObject(k);JSONObject p=l==null?null:l.optJSONArray("parents")==null?null:l.optJSONArray("parents").optJSONObject(0);if(p!=null&&"active".equalsIgnoreCase(p.optString("status","active"))&&!p.optString("phone","").trim().isEmpty()&&l.optBoolean("primary_guardian")){chosen=new Contact(displayName(s),p.optString("phone"),p.optString("name"),s.optString("id"));break;}}
                    if(chosen==null&&links!=null) for(int k=0;k<links.length();k++){JSONObject l=links.optJSONObject(k);JSONObject p=l==null?null:l.optJSONArray("parents")==null?null:l.optJSONArray("parents").optJSONObject(0);if(p!=null&&"active".equalsIgnoreCase(p.optString("status","active"))&&!p.optString("phone","").trim().isEmpty()){chosen=new Contact(displayName(s),p.optString("phone"),p.optString("name"),s.optString("id"));break;}}
                    if(chosen!=null)loaded.add(chosen);
                }
                synchronized(contacts){contacts.clear();contacts.addAll(loaded);}
                runOnUiThread(()->{refresh();summary.setText(contacts.size()+" active parent contacts loaded from ERP.");});
            }catch(Exception e){runOnUiThread(()->summary.setText("Could not load ERP parents: "+e.getMessage()));}
        }).start();
    }
    String displayName(JSONObject s){String n=(s.optString("first_name","")+" "+s.optString("middle_name","")+" "+s.optString("last_name","")).replaceAll("\\s+"," ").trim();return n+" ("+s.optString("admission_number","")+")";}
    void refresh(){
        if(list==null)return;list.removeAllViews();int selected=0;
        for(int i=0;i<contacts.size();i++){Contact c=contacts.get(i);LinearLayout row=new LinearLayout(this);row.setGravity(Gravity.CENTER_VERTICAL);
            CheckBox cb=new CheckBox(this);cb.setChecked(c.selected);cb.setText(c.student+" — "+c.parent+" — "+c.phone);cb.setOnClickListener(v->{c.selected=cb.isChecked();updateCounter();});row.addView(cb);list.addView(row);if(c.selected)selected++;}
        updateCounter();
    }
    void updateCounter(){int n=0;for(Contact c:contacts)if(c.selected)n++;counter.setText(n+" / 100 selected");}
    void loadSims(){try{SubscriptionManager sm=(SubscriptionManager)getSystemService(TELEPHONY_SUBSCRIPTION_SERVICE);if(Build.VERSION.SDK_INT>=22&&checkSelfPermission(Manifest.permission.READ_PHONE_STATE)==PackageManager.PERMISSION_GRANTED){List<SubscriptionInfo>x=sm.getActiveSubscriptionInfoList();if(x!=null)sims.addAll(x);}}catch(Exception ignored){}ArrayList<String> names=new ArrayList<>();for(int i=0;i<sims.size();i++)names.add("SIM "+(i+1)+" • "+sims.get(i).getCarrierName());if(names.isEmpty())names.add("Default SIM");simSpinner.setAdapter(new ArrayAdapter<String>(this,android.R.layout.simple_spinner_dropdown_item,names));}
    void startSending(){if(sending)return;String body=message.getText().toString().trim();ArrayList<Contact> selected=new ArrayList<>();for(Contact c:contacts)if(c.selected)selected.add(c);if(selected.isEmpty()){toast("Select parents from the ERP list.");return;}if(selected.size()>100){toast("Maximum 100 recipients per batch.");return;}if(body.isEmpty()){toast("Enter an SMS message.");return;}if(body.length()>160){toast("Keep SMS at 160 characters or less.");return;}sendBatch(body,selected);}
    void sendBatch(String body,ArrayList<Contact> selected){if(Build.VERSION.SDK_INT>=23&&checkSelfPermission(Manifest.permission.SEND_SMS)!=PackageManager.PERMISSION_GRANTED){requestPermissions(new String[]{Manifest.permission.SEND_SMS},REQ_SMS);return;}sending=true;sent=failed=delivered=0;total=selected.size();send.setEnabled(false);summary.setText("Sending 0 / "+total+"...");int pos=simSpinner.getSelectedItemPosition();Integer sub=(pos>=0&&pos<sims.size())?sims.get(pos).getSubscriptionId():null;for(int i=0;i<selected.size();i++){final int ix=i;Contact c=selected.get(i);handler.postDelayed(()->sendOne(c.phone,ix,sub),i*650L);}}
    void sendOne(String phone,int index,Integer subId){try{SmsManager mgr=subId!=null?SmsManager.getSmsManagerForSubscriptionId(subId):SmsManager.getDefault();Intent si=new Intent(this,SmsStatusReceiver.class).setAction("AIC_SMS_SENT").putExtra("index",index);Intent di=new Intent(this,SmsStatusReceiver.class).setAction("AIC_SMS_DELIVERED").putExtra("index",index);int flags=Build.VERSION.SDK_INT>=23?PendingIntent.FLAG_UPDATE_CURRENT|PendingIntent.FLAG_IMMUTABLE:PendingIntent.FLAG_UPDATE_CURRENT;mgr.sendTextMessage(phone,null,message.getText().toString().trim(),PendingIntent.getBroadcast(this,1000+index,si,flags),PendingIntent.getBroadcast(this,2000+index,di,flags));}catch(Exception e){failed++;updateSummary();}}
    void onSent(boolean ok,String phone){if(ok)sent++;else failed++;updateSummary();}
    void onDelivered(boolean ok){if(ok)delivered++;updateSummary();}
    void updateSummary(){runOnUiThread(()->{summary.setText("Sent: "+sent+"   Failed: "+failed+"   Delivered: "+delivered+"\\nProcessed: "+(sent+failed)+" / "+total);if(sent+failed>=total){sending=false;send.setEnabled(true);summary.setText("Finished • Sent "+sent+" • Failed "+failed+" • Delivered "+delivered+" / "+total);}});}
    JSONObject request(String url,String method,String body,String token)throws Exception{return new JSONObject(requestText(url,method,body,token));}
    String requestText(String url,String method,String body,String token)throws Exception{HttpURLConnection c=(HttpURLConnection)new URL(url).openConnection();c.setRequestMethod(method);c.setRequestProperty("apikey",SB_KEY);c.setRequestProperty("Content-Type","application/json");if(token!=null&&!token.isEmpty())c.setRequestProperty("Authorization","Bearer "+token);c.setDoInput(true);if(body!=null){c.setDoOutput(true);try(OutputStream o=c.getOutputStream()){o.write(body.getBytes("UTF-8"));}}int code=c.getResponseCode();InputStream in=code>=200&&code<300?c.getInputStream():c.getErrorStream();BufferedReader br=new BufferedReader(new InputStreamReader(in));StringBuilder sb=new StringBuilder();String line;while((line=br.readLine())!=null)sb.append(line);if(code<200||code>=300)throw new Exception("HTTP "+code+" "+sb);return sb.toString();}
    void toast(String s){Toast.makeText(this,s,Toast.LENGTH_SHORT).show();}
    static class Contact{String student,phone,parent,id;boolean selected;Contact(String s,String p,String pa,String i){student=s;phone=p;parent=pa;id=i;}}
    @Override protected void onDestroy(){super.onDestroy();if(instance==this)instance=null;}
}

File: android-messenger/app/src/main/java/ke/aiccathedral/schoolmessenger/MainActivity.kt
package ke.aiccathedral.schoolmessenger

import android.Manifest
import android.app.Activity
import android.app.PendingIntent
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.graphics.BitmapFactory
import android.graphics.Color
import android.telephony.SubscriptionInfo
import android.telephony.SubscriptionManager
import android.telephony.SmsManager
import android.widget.*
import org.json.JSONArray

class MainActivity : Activity() {
    private lateinit var connection: TextView
    private lateinit var queue: TextView
    private lateinit var messagePreview: TextView
    private lateinit var simSpinner: Spinner
    private lateinit var sendButton: Button
    private lateinit var progress: TextView
    private val handler = Handler(Looper.getMainLooper())
    private val sims = ArrayList<SubscriptionInfo>()
    private var pending = JSONArray()
    private var sending = false
    private var sent = 0
    private var failed = 0
    private var total = 0
    private val queueLoop = object : Runnable { override fun run() { if (!sending) refreshQueue(); handler.postDelayed(this, 3000) } }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        MainActivityHolder.activity = this
        window.statusBarColor = 0xFF7A1F3D.toInt()
        buildUi()
        requestPermissionsIfNeeded()
        loadSims()
        startService()
        refreshQueue()
        handler.postDelayed(queueLoop, 3000)
    }

    private fun buildUi() {
        val scroll = ScrollView(this)
        val root = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(28, 24, 28, 28)
        }
        scroll.addView(root)
        setContentView(scroll)

        val logo = ImageView(this).apply {
            setImageBitmap(BitmapFactory.decodeStream(resources.openRawResource(R.raw.aic_cathedral_logo)))
            adjustViewBounds = true
            scaleType = ImageView.ScaleType.FIT_CENTER
            setBackgroundColor(Color.WHITE)
            setPadding(0, 0, 0, 12)
            layoutParams = LinearLayout.LayoutParams(-1, 180)
        }
        root.addView(logo)

        val brand = TextView(this).apply {
            text = "AIC School Messenger"
            textSize = 28f
            setTextColor(0xFF7A1F3D.toInt())
            setTypeface(null, android.graphics.Typeface.BOLD)
            setPadding(0, 8, 0, 2)
        }
        root.addView(brand)
        root.addView(TextView(this).apply {
            text = "Automatic ERP SMS bridge"
            textSize = 16f
            setTextColor(0xFF666666.toInt())
            setPadding(0, 0, 0, 18)
        })

        connection = label("ERP connection: Checking…", 18f)
        root.addView(connection)

        queue = label("Waiting for an SMS from ERP…", 20f)
        queue.setTypeface(null, android.graphics.Typeface.BOLD)
        root.addView(queue)

        messagePreview = label("No pending message", 15f)
        messagePreview.setPadding(0, 4, 0, 18)
        root.addView(messagePreview)

        root.addView(label("Select sending SIM", 15f))
        simSpinner = Spinner(this)
        root.addView(simSpinner)

        sendButton = Button(this).apply {
            text = "SEND PENDING SMS"
            isEnabled = false
            setOnClickListener { sendPending() }
        }
        root.addView(sendButton)

        progress = label("Ready. Messages sent from the selected SIM.", 16f)
        progress.setPadding(0, 16, 0, 8)
        root.addView(progress)

        root.addView(Button(this).apply {
            text = "REFRESH ERP QUEUE"
            setOnClickListener { refreshQueue() }
        })

        root.addView(TextView(this).apply {
            text = "The app checks the ERP automatically. Keep this phone online with the school SIM active."
            textSize = 14f
            setTextColor(0xFF666666.toInt())
            setPadding(0, 18, 0, 0)
        })
    }

    private fun label(text: String, size: Float) = TextView(this).apply {
        this.text = text
        textSize = size
        setTextColor(0xFF242024.toInt())
        setPadding(0, 8, 0, 8)
    }

    private fun requestPermissionsIfNeeded() {
        val permissions = ArrayList<String>()
        if (Build.VERSION.SDK_INT >= 23 &&
            checkSelfPermission(Manifest.permission.SEND_SMS) != PackageManager.PERMISSION_GRANTED) {
            permissions.add(Manifest.permission.SEND_SMS)
        }
        if (Build.VERSION.SDK_INT >= 22 &&
            checkSelfPermission(Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_GRANTED) {
            permissions.add(Manifest.permission.READ_PHONE_STATE)
        }
        if (Build.VERSION.SDK_INT >= 33 &&
            checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
            permissions.add(Manifest.permission.POST_NOTIFICATIONS)
        }
        if (permissions.isNotEmpty()) requestPermissions(permissions.toTypedArray(), 77)
    }

    private fun loadSims() {
        sims.clear()
        try {
            if (Build.VERSION.SDK_INT >= 22 &&
                checkSelfPermission(Manifest.permission.READ_PHONE_STATE) == PackageManager.PERMISSION_GRANTED) {
                val manager = getSystemService(TELEPHONY_SUBSCRIPTION_SERVICE) as SubscriptionManager
                manager.activeSubscriptionInfoList?.let { sims.addAll(it) }
            }
        } catch (_: Exception) {}
        val names = ArrayList<String>()
        for (i in sims.indices) {
            val carrier = sims[i].carrierName?.toString()?.takeIf { it.isNotBlank() } ?: "Mobile network"
            names.add("SIM " + (i + 1) + " • " + carrier)
        }
        if (names.isEmpty()) names.add("Automatic / default SIM")
        simSpinner.adapter = ArrayAdapter(this, android.R.layout.simple_spinner_dropdown_item, names)
    }

    private fun startService() {
        val intent = Intent(this, SmsPollingService::class.java)
        if (Build.VERSION.SDK_INT >= 26) startForegroundService(intent) else startService(intent)
    }

    private fun refreshQueue() {
        Thread {
            try {
                val result = ApiClient.check(this)
                val recipients = result.optJSONArray("recipients") ?: JSONArray()
                runOnUiThread { applyQueue(result, recipients) }
            } catch (e: Exception) {
                runOnUiThread {
                    connection.text = "ERP connection: Not connected"
                    queue.text = "Waiting for ERP connection…"
                    messagePreview.text = e.message ?: "Connection error"
                    sendButton.isEnabled = false
                }
            }
        }.start()
    }

    private fun applyQueue(result: org.json.JSONObject, recipients: JSONArray) {
        pending = recipients
        connection.text = "ERP connection: Connected ✓"
        if (recipients.length() == 0) {
            queue.text = "No pending SMS"
            messagePreview.text = "When you send an SMS from ERP Communications, it will appear here automatically."
            sendButton.isEnabled = false
            return
        }
        val message = recipients.optJSONObject(0)?.optString("message", "") ?: ""
        queue.text = "${recipients.length()} SMS ready from ERP"
        messagePreview.text = if (message.length > 180) message.take(180) + "…" else message
        sendButton.isEnabled = !sending
    }

    private fun sendPending() {
        if (sending) return
        if (pending.length() == 0) {
            refreshQueue()
            return
        }
        if (Build.VERSION.SDK_INT >= 23 &&
            checkSelfPermission(Manifest.permission.SEND_SMS) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(arrayOf(Manifest.permission.SEND_SMS), 77)
            return
        }
        val selected = simSpinner.selectedItemPosition
        val subscriptionId = if (selected >= 0 && selected < sims.size) sims[selected].subscriptionId else null
        sending = true
        sent = 0
        failed = 0
        total = pending.length()
        sendButton.isEnabled = false
        progress.text = "Preparing ${total} SMS…"

        val ids = ArrayList<String>()
        for (i in 0 until pending.length()) {
            pending.optJSONObject(i)?.optString("id")?.takeIf { it.isNotBlank() }?.let { ids.add(it) }
        }

        Thread {
            try {
                val broadcastId = pending.optJSONObject(0)?.optString("broadcast_id", "") ?: ""
                ApiClient.claim(this, broadcastId, ids)
                for (i in 0 until pending.length()) {
                    val r = pending.optJSONObject(i) ?: continue
                    val id = r.optString("id")
                    val phone = r.optString("phone")
                    val body = r.optString("message")
                    handler.postDelayed({
                        sendOne(id, phone, body, i, subscriptionId)
                    }, i * 650L)
                }
            } catch (e: Exception) {
                runOnUiThread {
                    sending = false
                    sendButton.isEnabled = true
                    progress.text = "Could not start sending: ${e.message}"
                }
            }
        }.start()
    }

    private fun sendOne(id: String, phone: String, body: String, index: Int, subscriptionId: Int?) {
        try {
            val manager = if (subscriptionId != null && Build.VERSION.SDK_INT >= 22)
                SmsManager.getSmsManagerForSubscriptionId(subscriptionId)
            else SmsManager.getDefault()

            val sentIntent = Intent(this, SmsResultReceiver::class.java)
                .setAction("AIC_SMS_SENT")
                .putExtra("recipient_id", id)
                .putExtra("index", index)
                .putExtra("phone", phone)
                .setPackage(packageName)
            val deliveredIntent = Intent(this, SmsResultReceiver::class.java)
                .setAction("AIC_SMS_DELIVERED")
                .putExtra("recipient_id", id)
                .putExtra("index", index)
                .putExtra("phone", phone)
                .setPackage(packageName)
            val flags = if (Build.VERSION.SDK_INT >= 23)
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            else PendingIntent.FLAG_UPDATE_CURRENT

            manager.sendTextMessage(
                phone, null, body,
                PendingIntent.getBroadcast(this, 10000 + index, sentIntent, flags),
                PendingIntent.getBroadcast(this, 20000 + index, deliveredIntent, flags)
            )
        } catch (e: Exception) {
            onSendResult(false, id, index, "SMS error: ${e.message}")
        }
    }

    fun onSendResult(ok: Boolean, id: String, index: Int, error: String?) {
        if (ok) sent++ else failed++
        val processed = sent + failed
        runOnUiThread {
            progress.text = "Sent: ${sent}   Failed: ${failed}   Processed: ${processed} / ${total}"
            if (processed >= total) {
                sending = false
                sendButton.isEnabled = true
                progress.text = "Finished • Sent ${sent} • Failed ${failed} / ${total}"
                refreshQueue()
            }
        }
    }

    override fun onResume() {
        super.onResume()
        loadSims()
        refreshQueue()
    }

    override fun onDestroy() {
        handler.removeCallbacks(queueLoop)
        if (MainActivityHolder.activity === this) MainActivityHolder.activity = null
        super.onDestroy()
    }
}
